File: server/js/dialog.js
class Dialog {
  constructor(address, name) {
    this.address = address + '';
    this.name = name + '';
    this.lastBody = '';
    this.messages = [];
    this.lastDate = 0;
  }


  addSms(sms) {
    if (sms.date+0 > this.lastDate+0) {
      this.lastDate = sms.date;
      this.lastBody = sms.body;
    }
    this.messages.push(sms);
  }

  sort() {
    for (var i = 0; i < this.messages.length; i++)
      for (var j = i + 1; j < this.messages.length; j++)
        if (this.messages[i] > this.messages[j]) {
          var temp = this.messages[i];
          this.messages[i] = this.messages[j];
          this.messages[j] = temp;
        }
    this.body = this.messages[this.messages.length - 1]['body'];
  }

  isNotOneDay(first, second) {
      var one1 = new Date(first);
      var two1 = new Date(second);

      if (one1.getFullYear() == two1.getFullYear())
        if (one1.getMonth() == two1.getMonth())
          if (one1.getDate() == two1.getDate())
            return false;
      return true;
  }
}

File: smsmaket/app/src/main/java/com/alt/karman_pc/smsmaket/helperFiles/Dialog.kt
package com.alt.karman_pc.smsmaket.helperFiles

import me.everything.providers.android.telephony.Sms
import java.util.*

class Dialog (var address: String, var name: String, var avatar: String = "none") {

    var lastMessageDate: Long = 0
    var messages = emptyArray<RealmMessage>()
    var lastBody = ""

    fun addMessage(sms: RealmMessage) {
        if (messages.isEmpty()) {
            lastMessageDate = sms.getDate()
            lastBody = sms.body
        }
        messages += sms
    }

    fun search(text: String): Array<RealmMessage> {
        var result = emptyArray<RealmMessage>()
        for (sms in messages)
            if (sms.address != "date" && sms.body.contains(text, true))
                result += sms
        return result
    }

    fun generateStatistic(): Statistic {
        val countMessages: Int
        var countOutSms = 0
        var countInSms = 0
        var time = emptyArray<Long>()
        var chastota = emptyArray<Int>()


        for (sms in messages)
            if (sms.type != null) {
                if (sms.type == "INBOX")
                    countInSms++
                else if (sms.address != "date")
                    countOutSms++
            }

        var index = -1
        for (sms in messages) {
            if (sms.address == "date") {
                index++
                chastota += 0
                time += sms.sentDate
            }
            else
                try {
                    chastota[index]++
                } catch (e: Exception) {
                    index++
                    chastota += 0
                    time += sms.getDate()
                }
        }
        countMessages = countInSms + countOutSms
        return Statistic(countMessages, countInSms, countOutSms, chastota, time)
    }

    private fun isNotOneDay(first: Long, second: Long): Boolean {
        val one1 = Date(first)
        val two1 = Date(second)

        if (one1.year == two1.year
            && one1.month == two1.month
            && one1.day == two1.day)
            return false
        return true
    }

    private fun createDateSms(date: Long): RealmMessage {
        val dateSms = RealmMessage()
        dateSms.address = "date"
        dateSms.sentDate = date
        return dateSms
    }

    fun formatDialog() {
        messages.reverse()
        var lastDate: Long = 0
        var dialogArray = emptyArray<RealmMessage>()
        for (sms in messages) {
            if (isNotOneDay(sms.getDate(), lastDate) && Date(sms.getDate()).after(Date(lastDate))) {
                dialogArray += createDateSms(sms.getDate())
                lastDate = sms.getDate()
            }
            dialogArray += sms
        }
        messages = dialogArray
    }

    class Statistic(
        val allcount: Int,
        val inboxCount: Int,
        val outboxCount: Int,
        val chastota: Array<Int>,
        val time: Array<Long>
    ) {
        fun getProcInbox() = inboxCount.toFloat() / allcount
        fun getProcOutbox() = outboxCount.toFloat() / allcount
    }
}